cmd/qos.go

```diff
@@ -0,0 +1,74 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/pokt-network/poktroll/pkg/polylog"
+
+	"github.com/buildwithgrove/path/config"
+	"github.com/buildwithgrove/path/gateway"
+	"github.com/buildwithgrove/path/qos/evm"
+	"github.com/buildwithgrove/path/relayer"
+)
+
+// getServiceQoSInstances returns all QoS instances
+// to be used by the Gateway and EndpointHydrator, respectively.
+// This is done to ensure the same QoS instance is used in both
+// Gateway and EndpointHydrator, if the Service QoS implements
+// the gateway packages' QoSEndpointCheckGenerator interface.
+func getServiceQoSInstances(
+	gatewayConfig config.GatewayConfig,
+	logger polylog.Logger,
+) (
+	map[relayer.ServiceID]gateway.QoSService,
+	map[relayer.ServiceID]gateway.QoSEndpointCheckGenerator,
+	error,
+) {
+	// Build a mpa of services configured for the hydrator
+	// and the gateway to allow easy lookup.
+	hydratorServiceIDsIdx := buildServiceIDsIdx(gatewayConfig.HydratorConfig.ServiceIDs)
+	gatewayServiceIDsIdx := buildServiceIDsIdx(gatewayConfig.GetEnabledServiceIDs())
+
+	// TODO_TECHDEBT: refactor this function to remove
+	// the need for manually adding entries for every new QoS implmenetation.
+	gatewayQoSService := make(map[relayer.ServiceID]gateway.QoSService)
+	hydratorQoSGenerators := make(map[relayer.ServiceID]gateway.QoSEndpointCheckGenerator)
+
+	// TODO_FUTURE: support serviceQoS-specific configuration.
+	allServiceIDs := append(gatewayConfig.GetEnabledServiceIDs(), gatewayConfig.HydratorConfig.ServiceIDs...)
+	for _, serviceID := range allServiceIDs {
```

Comment:
Should we move this to its own file?

Reply:
I agree, but for the code being introduced in this PR I think this is sufficient.
We will likely split the code into more files once we address some of the TODOs here.